File: leetcode/python/1631_path_with_minimum_effort.py
import heapq


class Solution:
    def minimumEffortPath(self, heights: list[list[int]]) -> int:
        heap = [(0, (0, 0))]
        visited = set[tuple[int, int]]()
        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        m = len(heights)
        n = len(heights[0])
        goal = (m-1, n-1)
        while heap:
            (curMin, pos) = heapq.heappop(heap)
            if pos in visited:
                continue
            if pos == goal:
                return curMin
            
            visited.add(pos)
            for dy, dx in dirs:
                y = pos[0] + dy
                x = pos[1] + dx
                if 0 <= x < n and 0 <= y < m:
                    diff = abs(heights[pos[0]][pos[1]] - heights[y][x])
                    heapq.heappush(heap, (max(curMin, diff), (y, x)))

s = Solution()
print(s.minimumEffortPath([[1,2,2],[3,8,2],[5,3,5]]))
print(s.minimumEffortPath([[1,2,3],[3,8,4],[5,3,5]]))
print(s.minimumEffortPath([[1,2,1,1,1],[1,2,1,2,1],[1,2,1,2,1],[1,2,1,2,1],[1,1,1,2,1]]))

File: leetcode/rust/src/1631_path_with_minimum_effort.rs
use std::{collections::{HashSet, BinaryHeap}, cmp::Reverse};

impl Solution {
    pub fn minimum_effort_path(heights: Vec<Vec<i32>>) -> i32 {
        let m = heights.len();
        let n = heights[0].len();

        let mut heap = BinaryHeap::new();
        let mut visited = HashSet::new();

        heap.push(Reverse((0, 0, 0)));

        while !heap.is_empty() {
            let Reverse((effort, x, y)) = heap.pop().unwrap();
            if visited.contains(&(x, y)) {
                continue;
            }

            if x == m - 1 && y == n - 1 {
                return effort;
            }
            visited.insert((x, y));

            for (dx, dy) in [(0, 1), (0, -1), (1, 0), (-1, 0)].iter() {
                let nx = x as i32 + dx;
                let ny = y as i32 + dy;
                if nx < 0 || nx >= m as i32 || ny < 0 || ny >= n as i32 {
                    continue;
                }
                let nx = nx as usize;
                let ny = ny as usize;
                let new_effort = (heights[nx][ny] - heights[x][y]).abs().max(effort);
                heap.push(Reverse((new_effort, nx, ny)));
            }
        }

        0
    }
}